packages/analytics/amplify_analytics_pinpoint_dart/lib/src/impl/analytics_client/event_client/string_database/index_db/index_db_string_db.dart

```diff
@@ -0,0 +1,241 @@
+
+import 'dart:async';
+import 'dart:js_util';
+
+import 'package:amplify_analytics_pinpoint_dart/src/impl/analytics_client/event_client/string_database/index_db/in_memory_string_db.dart';
+import 'package:amplify_analytics_pinpoint_dart/src/impl/analytics_client/event_client/string_database/string_database.dart';
+import 'package:amplify_analytics_pinpoint_dart/src/impl/analytics_client/event_client/string_database/string_database_element.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:aws_common/src/js/indexed_db.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_analytics_pinpoint_dart.index_db_string_db}
+/// Stores string elements using [InMemoryStringDb] and IndexDB for persistence
+/// {@endtemplate}
+class IndexDbStringDb
+    with AWSDebuggable, AWSLoggerMixin
+    implements StringDatabase {
+  IndexDbStringDb._(this._isIndexDbSupported, this._inMemoryDb, this._indexDb) {
+    if (!_isIndexDbSupported) {
+      logger.warn(
+        'IndexedDB is not available. '
+        'Falling back to in-memory storage.',
+      );
+    }
+  }
+
+  final bool _isIndexDbSupported;
+  final InMemoryStringDb _inMemoryDb;
+  final _IndexedDBStorage? _indexDb;
+
+  /// Handle edge case where we don't know IndexDB lastID if it's empty
+  static Future<int> _calculateLastId(
+    Iterable<StringDatabaseElement> readElements,
+    _IndexedDBStorage indexDb,
+  ) async {
+    if (readElements.isEmpty) {
+      await indexDb.addString('test');
+      final currentId = (await indexDb.getStrings(1)).first.id;
+      await indexDb.deleteStrings([currentId]);
+      return currentId + 1;
+    } else {
+      return readElements.last.id;
+    }
+  }
+
+  /// {@macro amplify_analytics_pinpoint_dart.index_db_string_db}
+  static Future<IndexDbStringDb> getInstance() async {
+    final isIndexDbSupported = await _checkIsIndexDbSupported();
+
+    if (isIndexDbSupported) {
+      final indexDb = _IndexedDBStorage();
+      final readElements = await indexDb.getAllStrings();
+
+      final lastId = await _calculateLastId(readElements, indexDb);
+      final inMemoryDb =
+          InMemoryStringDb(initValues: readElements.toList(), lastId: lastId);
+      return IndexDbStringDb._(isIndexDbSupported, inMemoryDb, indexDb);
+    } else {
+      return IndexDbStringDb._(isIndexDbSupported, InMemoryStringDb(), null);
+    }
+  }
+
+  static Future<bool> _checkIsIndexDbSupported() async {
+    if (indexedDB == null) {
+      return false;
+    }
+    // indexedDB will be non-null in Firefox private browsing,
+    // but will fail to open.
+    try {
+      final openRequest = indexedDB!.open('test', 1);
+      await openRequest.future;
+      return true;
+    } on Exception {
+      return false;
+    }
+  }
+
+  @override
+  String get runtimeTypeName => 'IndexDbStringDb';
+
+  @override
+  FutureOr<void> addString(String string) async {
+    _inMemoryDb.addString(string);
+
+    try {
+      return _indexDb?.addString(string);
```

Comment:
`addString` needs to be awaited for the try/catch to work. Applies to all try/catches throughout where the body of `try` is a future.